File: Data Structures & Algorithms/daily-temperatures/submission-3.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> nextDay(n);
        vector<int> ans(n);
        nextDay[n - 1] = -1;

        for(int i = n - 2; i >= 0; i--) {
            int j = i + 1;
            while(j != -1 && j < n && temperatures[j] <= temperatures[i]) {
                j = nextDay[j];
            }

            nextDay[i] = j;
        }

        for(int i = 0; i < n; i++) {
            if(nextDay[i] == -1) {
                ans[i] = 0;
            } else {
                ans[i] = nextDay[i] - i;
            }
        }

        return ans;

    }
};

File: Data Structures & Algorithms/non-cyclical-number/submission-3.py
class Solution:
    def isHappy(self, n: int) -> bool:
        visited = set()

        while n not in visited:
            visited.add(n)
            sum_ = self.getSumSq(n)
            if(sum_ == 1):
                return True
            
            n = sum_
        return False

    def getSumSq(self, n: int) -> int:
        ans = 0
        while n != 0:
            ans += (n % 10) * (n % 10)
            n = n // 10
        return ans